### ISSUE TITLE
Unable to List Blobs in Azure via S3Proxy

### ISSUE THREAD

**Issue by samuel-davis** [gaul/s3proxy#717]
Problem: 
Utilizing S3Proxy configured to talk to ABS I cannot use the `marker` parameter when I attempt to ListBlobs.  I am using a underlying python library in my codebase called `deltalake` , where the `deltalake` lib is calling S3Proxy. 

The incoming request to S3Proxy from my code is : 

```
http://df-s3-proxy/df-bucket?list-type=2&prefix=av2%2Fsilver8%2F_delta_log%2F&start-after=av2%2Fsilver8%2F_delta_log%2F00000000000000000000
```
The transformed request that is sent JCloud and then on to ABS in the cloud is :
```
https://vulcanforgetest.blob.core.windows.net/df-bucket?restype=container&comp=list&prefix=av2/silver8/_delta_log/&marker=av2/silver8/_delta_log/00000000000000000000&maxresults=1000&include=metadata
``` 

Finally the exception caused by the marker parameter is :

```
org.jclouds.azure.storage.AzureStorageResponseException: command [method=org.jclouds.azureblob.AzureBlobClient.public abstract org.jclouds.azureblob.domain.ListBlobsResponse org.jclouds.azureblob.AzureBlobClient.listBlobs(java.lang.String,org.jclouds.azureblob.options.ListBlobsOptions[])[df-bucket, [Lorg.jclouds.azureblob.options.ListBlobsOptions;@41358ef0], request=GET https://vulcanforgetest.blob.core.windows.net/df-bucket?restype=container&comp=list&prefix=av2/silver8/_delta_log/&marker=av2/silver8/_delta_log/00000000000000000000&maxresults=1000&include=metadata HTTP/1.1] failed with code 400, error: AzureError{requestId='9606be37-201e-0072-4006-3223a9000000', code='InvalidQueryParameterValue', message='Value for one of the query parameters specified in the request URI is invalid.
RequestId:9606be37-201e-0072-4006-3223a9000000
Time:2024-11-08T17:46:11.9123328Z', context='{QueryParameterValue=av2/silver8/_delta_log/00000000000000000000, QueryParameterName=marker, Reason=Invalid ListBlobs marker.}'}
``` 

You can see above that it basically boils down to this. 

![image](https://github.com/user-attachments/assets/e9b92049-ef3a-4c13-8061-deda270448fe)

Ive looked through both the S3Proxy and the Jclouds code and because the error is so abstract and doesn't tell me why this marker parameter is invalid, Im reaching out for some help.  

Im more than happy to do a PR if you can point to where this can be resolved. 

I should also say, I attempted using the `azureblob-sdk` provider and while it DOES write data and gets past this error, the error that is written is unable to be read correctly afterwards which is implying that its not writing correctly. 

Environment where failure is seen: 
Azure Blob Storage account configured with : `azureSharedKey` in S3Proxy
provider: azureblob

**Comment by gaul** [comment]
I believe that your S3 client is using a marker that was not supplied by the object store.  S3 supports any arbitrary string as a marker, for example:

S3 has objects [a, b, d, f]
list maxResults = 1, marker = a -> returns b
list maxResults = 1, marker = b -> returns d
list maxResults = 1, marker = c -> returns d

However Azure has [opaque markers](https://learn.microsoft.com/en-us/rest/api/storageservices/enumerating-blob-resources#Subheading2) which are not simple strings.  Thus if you issue the same set of operations:

Azure has objects [a, b, d, f]
list maxResults = 1, marker = a -> returns b
list maxResults = 1, marker = b -> returns d
list maxResults = 1, marker = c -> emits error

How have you configured your client such that it is using an unexpected marker?  Sometimes clients use the last key instead of the marker while listing but S3Proxy has some fixup logic for this.  Also some clients try to list large buckets in parallel using random keys but S3Proxy cannot support this for either `azureblob` or `azureblob-sdk`.

Please open a separate issue for the azureblob error.  While it will likely have the same marker limitation, I am more likely to fix something if data is not being written properly.  Note that many changes have landed to it in the last for weeks so please ensure you use the latest version.

**Comment by samuel-davis** [comment]
Thank you for the explanation. 

In your example above about opaque markers in Azure , this particular example stands out. 

`list maxResults = 1, marker = c -> emits error` 

Deltalake is attempting to determine what number to supply to the log before it writes data . I believe it is using 0000000000000 so it can get a list and then use the highest number. 

However in this case, on the first write operation, there is nothing in the directory at all yet, so a invalid marker error is thrown. Does this line up with your explanation ? 

As for the separate issue, I'll write it up using the azureblob provider with Azurite as a emulator ( works perfectly there since azurite treats all invalid markers as a empty string) . Get the file sizes that were written and then compare to the azurblob-sdk provider with the same emulator , and then the difference when pointing towards the real Azure Blob Storage.

**Comment by gaul** [comment]
I don't mean that your code is incorrect but that the underlying S3 library may be doing an operation that S3Proxy using Azure cannot support.  Can you share specifically which library you use and how you call it?

**Comment by gaul** [comment]
Two of the three things tangled together in this thread have moved.

@jdunham-openai's case is fixed on master in 25d40a7. The frontend keeps a map
from the last key of a listing back to the marker that produced it, for stores
whose marker is a token rather than a key -- but it filed each entry under the
marker instead of under the key, so the only lookup that could ever hit was one
for a marker, which needs no lookup at all since an unrecognised value passes
through untouched. Paging by the last key, which S3 allows, reached Azure
verbatim and was refused. #569 had the cause right; the fix files the entry
under the key, which also covers ListObjectsV2 start-after.

@samuel-davis's original case is a different problem and is not fixed by that. A
start-after naming a key that no listing ever returned -- `...00000000000000000000`
against a directory that is empty on first write -- cannot be translated into a
marker, because there is nothing to translate it from. Azure's List Blobs had no
way to begin at an arbitrary name at all.

That is what changed. Service version 2026-02-06 adds a `startFrom` parameter
that does what S3's start-after does, natively. #938 tracks adopting it, and
adopting it removes the marker translation rather than extending it. The
remaining blocker was Azurite, which now has a PR: Azure/Azurite#2688.

@samuel-davis, separately you reported that azureblob-sdk wrote data that read
back smaller than it should be. That is a different bug and has been getting
lost here. The multipart tests now compare full content against the Azurite lane
and pass, so if you still see it, could you open it as its own issue with a
reproducer?

**Comment by gaul** [comment]
#938 is now fixed.  Please open new issues for separate symptoms.
